hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java

```diff
@@ -281,16 +298,43 @@ private static boolean applyConstraints(
 						descriptor.getComposingConstraints(),
 						property, propertyDesc, null,
 						canApplyNotNull,
+						isConstraintCompositionOfTypeOr( descriptor ),
 						dialect
 				);
 
-				hasNotNull = hasNotNull || hasNotNullFromComposingConstraints;
+				hasNotNull = compositionLogic.apply( hasNotNull, hasNotNullFromComposingConstraints );
 			}
-
+		}
+		if ( hasNotNull ) {
+			markNotNull( property );
 		}
 		return hasNotNull;
 	}
 
+	private static boolean isConstraintCompositionOfTypeOr(final ConstraintDescriptor<?> descriptor) {
+		if ( descriptor.getComposingConstraints() == null ||
+				descriptor.getComposingConstraints().size() < 2 ) {
+			return false;
+		}
+
+		return CONSTRAINT_COMPOSITION_TYPE_CACHE.computeIfAbsent( descriptor.getAnnotation().getClass(), unused -> {
+			// This check assumes that Hibernate Validator is being used providing an
+			// implementation that gives us the composition type that is being used.
+			try {
+				Method compositionTypeMethod = descriptor.getClass()
+						.getMethod( "getCompositionType" );
```

Comment:
Wait, is this calling a method of the internal class `ConstraintDescriptorImpl`?

I was expecting this to work be reflecting on `ConstraintComposition` annotations.

Reply:
Yes, did not think of this tbh. Makes more sense to reflect the annotations present. Will update accordingly.